rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,657 @@
+
+To ensure that the updater is always able to retrieve the desired version, instructions to the updater are delivered via unauthenticated requests to `/v1/webapi/find`.
+Teleport auth servers use their access to heartbeat data to drive the rollout, while Teleport proxies modulate the `/v1/webapi/find` response given the host UUID.
+
+Rollouts are specified as interdependent groups of hosts, selected by upgrade group identifier.
+```
+teleport:
+  upgrade_group: staging
+```
+
+At the start of a group rollout, the Teleport auth server captures the desired group of hosts to update in the backend.
+An fixed number of hosts (`max_in_flight % x total`) are instructed to upgrade at the same time via `/v1/webapi/find`.
+Additional hosts are instructed to update as earlier updates complete, timeout, or fail, never exceeding `max_in_flight`.
+The group rollout is halted if timeouts or failures exceed their specified thresholds.
+Group rollouts may be retried with `tctl autoupdate run`.
+
+### Scalability
+
+#### Window Capture
+
+Instance heartbeats will be cached by auth servers using a dedicated cache.
+This cache is updated using rate-limited backend reads that occur in the background, to avoid mass-reads of instance heartbeats.
+The rate is modulated by the total number of instance heartbeats, to avoid putting too much load on the backend on large clusters.
+The cache is considered healthy when all instance heartbeats present on the backend have been read in a time period that is also modulated by the total number of heartbeats.
+
+At the start of the upgrade window, auth servers attempt to write an update rollout plan to the backend under a single key.
+This plan is protected by optimistic locking, and contains the following data:
+
+Data key: `/autoupdate/[name of group]/[scheduled type](/[page-id])` (e.g., `/autoupdate/staging/critical/8745823`)
+
+Data value JSON:
+- `start_time`: timestamp of current window start time
+- `version`: version for which this rollout is valid
+- `hosts`: list of host UUIDs in randomized order
+- `next_page`: additional UUIDs, if list is greater than 100,000 UUIDs
+- `expiry`: 2 weeks
+
+At a fixed interval, auth servers will check the plan to determine if a new plan is needed by comparing `start_time` to the current time and the desired window.
+If a new plan is needed, auth servers will query their cache of instance heartbeats and attempt to write the new plan.
+The first auth server to write the plan wins; others will be rejected by the optimistic lock.
+Auth servers will only write the plan if their instance heartbeat cache is healthy.
+
+If the list is greater than 100,000 UUIDs, auth servers will first write pages with a randomly generated suffix, in a linked-link, before the atomic non-suffixed write.
+If the non-suffixed write fails, the auth server is responsible for cleaning up the unusable pages.
+If cleanup fails, the unusable pages will expire after 2 weeks.
```

Comment:
Any size limit based on the number of host IDs is likely to be wrong since we allow custom user supplied host IDs of any format.  Additionally, we should avoid creating backend values over 100kb in total size, so even if we were certain that they were uuids we couldn't store 100k in one value.  (note: 100kb is arbitrary, but the true hard limit is in that ballpark... couple hundred kb at absolute most... best not to push it)

I would suggest instead picking a maximum encoded length of 100kb and dynamically building the pages with size based on the size of the specific host IDs in question

Reply:
Fixed

Reply:
We could store (uniformly sized) hashes instead, with the only drawback being that getting back a list of host ids requires getting the full list of known hosts and back referencing them.